lib/spack/spack/solver/concretize.lp

```diff
@@ -137,22 +137,21 @@ attr(Name, Arg1, Arg2, Arg3) :-
 %-----------------------------------------------------------------------------
 % Virtual dependencies
 %-----------------------------------------------------------------------------
-% if you declare a dependency on a virtual AND the package is not an external,
-% you depend on one of its providers
-1 {
-  depends_on(Package, Provider, Type) : possible_provider(Provider, Virtual)
```

Comment:
This is the gist of the issue: we shouldn't use `possible_provider` since we don't know if the possible provider is **actually** a provider. For some reasons in #20679 `libiconv` was selected as the `iconv` provider to satisfy the link dependency, while `libc` was selected to satisfy the build dependency without being a provider of anything (but being a possible provider of `iconv`).

Reply:
For the record: I spent a couple of hours trying to come up with a reproducer for tests, but I was not able to find one (while I am able to reproduce the issue reported in #20679 on the command line).